### DISCUSSION TITLE
Catapult assisted takeoff - launch phase settings

### DISCUSSION THREAD

**Discussion by tpleece** [ImperialCollegeLondon/sharpy#197] [Q&A] [answered]
Hi,

I'm working an a HALE project using SHARPy and I wish to perform some dynamic coupled modelling of launch/ catapult-assisted takeoffs. I'm using the generate_hale.py file in sharpy/cases/coupled/multibody_takeoff as a reference but I can't see how the launch process is implemented in this file, it seems to go straight to free flight. Can you direct me to any documentation on the generate_dyn_file function and how to use the dynamic_forces and forced_for_vel capabilities?

**Comment by ngoiz** [comment, marked as answer]
Hi @tpleece,

This is indeed a complex case

For the catapult-takeoff likely the best case to start off with is [this one](https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/base_case/generate.py) and you can see in the function below that the takeoff settings are specified through Lagrange 
https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/base_case/generate.py#L717
and then implemented through [controllers](https://ic-sharpy.readthedocs.io/en/master/includes/controllers/takeofftrajectorycontroller/TakeOffTrajectoryController.html) https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/base_case/generate.py#L886

This is not my area of expertise but hopefully from this optimisation case (the [generate.py](https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/base_case/generate.py) file is used within the optimisation routine  [here](https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/optimiser.py))

Hope this helps,
Norberto

**Reply by tpleece** [reply]
Hi Norberto,

Thanks for your help. I'm having some difficulties with the case from the generate.py script you suggested. (First I had to remove a small number of solver settings that SHARPy didn't recognise - I assume that's just due to updated version of SHARPy) SHARPy seems to only solve for the initial equilibrium shape of the aircraft wing using the dynamic solver until the end of the simulation; the launch stage never starts. I've tried changing various settings in the DynamicMultibody solver, generate_trajectory_file and generate_multibody_file functions but can't get it to work. Is there anything else required to initiate the launch stage? 

Thanks,

Thomas

**Reply by ngoiz** [reply]
Hi Thomas,

First, yes, the settings that had to be removed is a new feature of SHARPy to avoid old settings creeping along (or typos etc).

I've had another look at the script and the only thing I can think of is that you don't have the `StaticTrim` solver in the flow, as otherwise the trajectory file will not be generated. That is my only guess if your equilibrium shape is being computed. If you attach the output log maybe we can have another look.

Thanks,
Norberto

**Reply by tpleece** [reply]
Hi Norberto,

The trajectory file was being generated previously. Line 979 ` if 'StaticTrim' not in flow:
        generate_trajectory_file()` suggests I should not be using `StaticTrim`. If I remove the if statement and add `StaticTrim` to the flow, I have to set initial thrust to 0 to get a converged trim condition. But I still get the same result of just solving for an equilibrium shape with the launch stage never starting. 
I've attached my output.log, if you are able to take a look that would be really appreciated.
[output.log](https://github.com/ImperialCollegeLondon/sharpy/files/8891849/output.log)

        
Thanks,
Thomas

**Reply by ngoiz** [reply]
Hi Thomas,

Hmm, the simulation log looks alright to me. Despite of the `StaticTrim` setting the thrust to 0, I would have expected the launch to initiate and then if the thrust is 0 the plane to glide, but indeed the aircraft remains static at the initial condition.

By any chance, can you try without `StaticTrim` in the `flow`? i.e. just run this:
```    
flow = ['BeamLoader',
            'AerogridLoader',
            'StaticCoupled',
            'DynamicCoupled',
            ]
```
Maybe give the aircraft some non-zero thrust but I wouldn't think that should affect the launch phase.

Then, hopefully the columns in the trajectory file do make sense to you, i.e. they should be positions in x and z, which will be a function of the ramp angle, 
https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/scripts/optimiser/base_case/generate.py#L282
so if they are correct that's the only thing I can think of.

Let's see if the trim issue is what is holding it back, if not, the final thing to test out is whether we have broken something in the development phase, as the CATO optimisation was done around 2020. If nothing else works, I would give it a shot with [v1.1.1](https://github.com/imperialcollegelondon/sharpy/tree/1.1.1)

Hope this helps,
Norberto

Or maybe @fonsocarre has some insight into how to get the launch to work.

(Modified the title for future reference)

**Reply by tpleece** [reply]
Hi Norberto,

Thanks again for your help. I tried again without `StaticTrim` in the flow and it still did not work. The Trajectory file makes sense to me and is being generated correctly. 

I've now got v1.1.1 running and now the launch phase is working properly.
[base.log](https://github.com/ImperialCollegeLondon/sharpy/files/8918693/base.log)
So it does seem like something has broken in the more recent versions, I'll have a look around to see if I can identify where and try to help fix it if I can. Should I open a separate bug report for this? 

Thanks,

Thomas

**Reply by ngoiz** [reply]
Many thanks, Thomas, for trying these options out, I guess the "good news" is that we now have something to look for.

I'll transform the issue into a bug.

I was having a quick look and the `TakeoffTrajectoryController` has seen no changes since v1.1.1, so I'm guessing the issue is in the `DynamicCoupled` solver that new routines may be overwriting the output from the controller
https://github.com/ImperialCollegeLondon/sharpy/blob/41b6cb699fe33167d66cbe0c5ede1bdfaaeb50c4/sharpy/solvers/dynamiccoupled.py#L501

I'll see if I have some time and have a more thorough look as well.

Thanks,
Norberto
